test/nlp.jl

```diff
@@ -328,6 +337,67 @@
         @test h ≈ correct_hessian*v
     end
 
+    @testset "Jacobians and Jac-vec" begin
+        m = Model()
+        @variable(m, a)
+        @variable(m, b)
+        @variable(m, c)
+
+        @NLobjective(m, Min, a*b + c^2)
+        @NLconstraint(m, c*b <= 1)
+        @NLconstraint(m, a^2/2 <= 1)
+        d = JuMP.NLPEvaluator(m)
+        MOI.initialize(d, [:JacVec, :Jac])
+
+        values = [1.0, 2.0, 3.0] # For a, b, c.
+        jacobian_sparsity = MOI.jacobian_structure(d)
+        V = zeros(length(jacobian_sparsity))
+        MOI.eval_constraint_jacobian(d, V, values)
+        correct_jacobian = [0.0 3.0 2.0; 1.0 0.0 0.0]
+        @test isapprox(dense_jacobian(jacobian_sparsity, V, 2, 3), correct_jacobian)
+
+        v = [2.4, 3.5, 1.2]
+        Jv = zeros(2)
+        MOI.eval_constraint_jacobian_product(d, Jv, values, v)
+        @test isapprox(Jv, correct_jacobian * v)
+
+        w = [0.6, 4.3]
+        Jᵀw = zeros(3)
```

Comment:
The ᵀ symbol could potentially confuse a reader. It's not obvious that it's part of the name and not an arithmetic operator. I would just re-use the name `product_storage`.